### PR TITLE
Refactor navigation links from left menu hard code to workspace plugin register

### PR DESCRIPTION
**Pull request by yuye-aws**
### Description

1. Check whether workspace is enabled for left menu.
2. Refactor to move workspace related nav link from left menu hardcode to workspace plugin.
3. Move exit workspace button from left menu to workspace update page.
4. Add nav links: "Home", "See more", "Overview" and "Admin" to the left menu. 
5. Recover recently viewed items for the left menu.

## Screenshot

Inside workspace:
image
 
Outside workspace:
image

Workspace disabled:
image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/public/chrome/ui/header/collapsible_nav.test.tsx [hunk 1]
```diff
@@ -166,12 +167,14 @@ describe('CollapsibleNav', () => {
...
     ];
     const recentNavLinks = [mockRecentNavLink({})];
+    const customNavLink = mockLink({ title: 'Custom link' });
     const component = mount(
       <CollapsibleNav
         {...mockProps()}
         isNavOpen={true}
         navLinks$={new BehaviorSubject(navLinks)}
         recentlyAccessed$={new BehaviorSubject(recentNavLinks)}
+        customNavLink$={new BehaviorSubject(customNavLink)}
```
Comment (wanglam) [inline src/core/public/chrome/ui/header/collapsible_nav.test.tsx:177]: What about add this customNavLink$ to mockProps function?

File: src/core/public/chrome/ui/header/collapsible_nav.tsx [hunk 2]
```diff
@@ -103,10 +101,6 @@ interface Props {
   navigateToUrl: InternalApplicationStart['navigateToUrl'];
   customNavLink$: Rx.Observable<ChromeNavLink | undefined>;
```
Comment (wanglam) [inline src/core/public/chrome/ui/header/collapsible_nav.tsx:102]: The `customNavLink$` is not used anymore. Do we need to add its logic back?

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx [hunk 3]
```diff
@@ -129,7 +127,54 @@ export const WorkspaceUpdater = () => {
       }
     }
     setDeleteWorkspaceModalVisible(false);
-    await application.navigateToApp('home');
+    window.location.href =
+      workspaces?.formatUrlWithWorkspaceId(
```
Comment (wanglam) [inline src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:131]: I can't get the point here. Why do we need to generate a workspace link with empty workspace id? How about just navigate to 
```js
application.getUrlForApp('home', {
          path: '/',
          absolute: true,
        })
```

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx [hunk 4]
```diff
@@ -129,7 +127,54 @@ export const WorkspaceUpdater = () => {
...
+      notifications?.toasts.addDanger({
+        title: i18n.translate('workspace.exit.failed', {
+          defaultMessage: 'Failed to exit workspace',
+        }),
+        text: result?.error,
+      });
+      return;
+    }
+    window.location.href =
+      workspaces?.formatUrlWithWorkspaceId(
```
Comment (wanglam) [inline src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:167]: The same as deleteWorkspace. Is it better to just redirect to home page?

File: src/plugins/workspace/public/plugin.ts [hunk 5]
```diff
@@ -106,9 +139,28 @@ export class WorkspacesPlugin implements Plugin<{}, {}, WorkspacesPluginSetupDep
...
+  ) {
+    if (!workspace) return allNavLinks;
+    const features = workspace.features ?? [];
+    return allNavLinks.filter((item) => features.includes(item.id));
+  }
+
   public start(core: CoreStart) {
     // If workspace feature is disabled, it will not load the workspace plugin
     if (core.uiSettings.get('workspace:enabled') === false) {
+      // set filtered nav links
```
Comment (wanglam) [inline src/plugins/workspace/public/plugin.ts:154]: How about move these default nav links setup logic to core plugin? I think these shouldn't be a part of workspace plugin?

File: src/core/public/chrome/ui/header/collapsible_nav.tsx (hunk 2 above)
Comment (yuye-aws) [reply src/core/public/chrome/ui/header/collapsible_nav.tsx:102]: Maybe we need this props in the future. What about just keep it currently?

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx (hunk 3 above)
Comment (yuye-aws) [reply src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:131]: Exiting workspace through `application.navigateToUrl` would result in wrong target url like:
`http://localhost:5601/w/5786ccb0-26a7-11ee-8cfa-113158c5ccb1/app/home#/`
This is because the workspace id will be part of the url when inside workspace. Do you have any suggestions?

Comment (wanglam) [reply src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:131]: Consider use `src/core/public/http/base_path.ts`, combine `remove` and `prepand` without workspace?

**Review by wanglam — APPROVED** [review]

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx [hunk 6]
```diff
@@ -129,7 +127,50 @@ export const WorkspaceUpdater = () => {
       }
     }
     setDeleteWorkspaceModalVisible(false);
-    await application.navigateToApp('home');
+    if (http) {
+      await application.navigateToUrl(
+        http.basePath.remove(
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:132]: `workspace.formatUrlWithWorkspaceId` can remove the workspace prefix (/w/${workspaceId}) if passed `workspaceId` is empty, basePath.remove will remove the serverBasePath as well, which will lead to bug.

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx [hunk 7]
```diff
@@ -129,7 +127,50 @@ export const WorkspaceUpdater = () => {
...
+      notifications?.toasts.addDanger({
+        title: i18n.translate('workspace.exit.failed', {
+          defaultMessage: 'Failed to exit workspace',
+        }),
+        text: result?.error,
+      });
+      return;
+    }
+    if (http) {
+      await application.navigateToUrl(
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:165]: Same here, `workspace.formatUrlWithWorkspaceId` can remove the workspace prefix (/w/${workspaceId}) if passed `workspaceId` is empty, basePath.remove will remove the serverBasePath as well, which will lead to bug.

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx [hunk 8]
```diff
@@ -139,6 +180,7 @@ export const WorkspaceUpdater = () => {
           restrictWidth
           pageTitle="Update Workspace"
           rightSideItems={[
+            <EuiButton onClick={() => exitWorkspace()}>Exit</EuiButton>,
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:183]:
```suggestion
            <EuiButton onClick={exitWorkspace}>Exit</EuiButton>,
```
exitWorkspace, do not create a new arrow function here.

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx (hunk 6 above)
Comment (yuye-aws) [reply src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:132]: By specifying `absolute` to false, the `getUrlForApp` will not include serverBasePath, which will not be removed.
```
          application.getUrlForApp('home', {
            path: '',
            absolute: false,
          })
```

File: src/core/public/chrome/ui/header/collapsible_nav.test.tsx [hunk 9]
```diff
@@ -143,14 +138,12 @@ describe('CollapsibleNav', () => {
...
       mockRecentNavLink({ label: 'recent 2' }),
     ];
-    const customNavLink = mockLink({ title: 'Custom link' });
     const component = mount(
       <CollapsibleNav
         {...mockProps()}
         isNavOpen={true}
         navLinks$={new BehaviorSubject(navLinks)}
         recentlyAccessed$={new BehaviorSubject(recentNavLinks)}
-        customNavLink$={new BehaviorSubject(customNavLink)}
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav.test.tsx:153]: This line should not be deleted I think?

File: src/core/public/chrome/ui/header/collapsible_nav.test.tsx [hunk 10]
```diff
@@ -143,14 +138,12 @@ describe('CollapsibleNav', () => {
       mockRecentNavLink({ label: 'recent 1' }),
       mockRecentNavLink({ label: 'recent 2' }),
     ];
-    const customNavLink = mockLink({ title: 'Custom link' });
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav.test.tsx:146]: This line should not be deleted I think?

File: src/core/public/chrome/ui/header/collapsible_nav.test.tsx (hunk 9 above)
Comment (yuye-aws) [reply src/core/public/chrome/ui/header/collapsible_nav.test.tsx:153]: `customNavLink$` is defined in mockProps

File: src/core/public/chrome/ui/header/collapsible_nav.test.tsx [hunk 11]
```diff
@@ -75,16 +74,12 @@ function mockProps() {
     homeHref: '/',
     navLinks$: new BehaviorSubject([]),
     recentlyAccessed$: new BehaviorSubject([]),
+    customNavLink$: new BehaviorSubject(mockLink({ title: 'Custom link' })),
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/collapsible_nav.test.tsx:77]: I do not know why we change the code here, we'd better leave it untouched.

File: src/core/public/chrome/ui/header/nav_link.tsx [hunk 12]
```diff
@@ -31,14 +31,9 @@
...
-import {
-  ChromeNavLink,
-  ChromeRecentlyAccessedHistoryItem,
-  CoreStart,
-  WorkspaceAttribute,
-} from '../../..';
+import { ChromeNavLink, ChromeRecentlyAccessedHistoryItem, CoreStart } from '../../..';
 import { HttpStart } from '../../../http';
-import { InternalApplicationStart } from '../../../application/types';
+import { InternalApplicationStart } from '../../../application';
```
Comment (SuZhou-Joe) [inline src/core/public/chrome/ui/header/nav_link.tsx:36]: Unnecessary change, please revert this line, it may be the same on function but may result in unnecessary code conflict when we merge our code back to OSD main/2.x .

File: src/core/public/core_system.ts [hunk 13]
```diff
@@ -233,7 +233,17 @@ export class CoreSystem {
         injectedMetadata,
         notifications,
         uiSettings,
-        workspaces,
+      });
+
+      // set default value for filtered nav links
+      const navLinksService = chrome.navLinks;
```
Comment (SuZhou-Joe) [inline src/core/public/core_system.ts:239]: From the code, the default value will be all the links that not filtered by featureGrup/feature flag/current workspace. I would suggest not adding the default value here, as it will trigger a unnecessary render of collapse_nav component.

File: src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx (hunk 6 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/components/workspace_updater/workspace_updater.tsx:132]: https://github.com/ruanyl/OpenSearch-Dashboards/assets/13493605/df778617-235e-4ccb-9f84-a51a932d89d3




With `server.basePath` and `server.rewriteBasePath` enabled, it jumps to 404 after deleting a workspace.

File: src/plugins/workspace/public/plugin.ts [hunk 14]
```diff
@@ -120,6 +172,96 @@ export class WorkspacesPlugin implements Plugin<{}, {}, WorkspacesPluginSetupDep
       chrome: core.chrome,
     });
     this.currentWorkspaceSubscription = this._changeSavedObjectCurrentWorkspace();
+    if (core) {
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/plugin.ts:175]: Can we move this huge block to a private method?

File: src/plugins/workspace/public/plugin.ts [hunk 15]
```diff
@@ -106,9 +139,28 @@ export class WorkspacesPlugin implements Plugin<{}, {}, WorkspacesPluginSetupDep
...
+  ) {
+    if (!workspace) return allNavLinks;
+    const features = workspace.features ?? [];
+    return allNavLinks.filter((item) => features.includes(item.id));
+  }
+
   public start(core: CoreStart) {
     // If workspace feature is disabled, it will not load the workspace plugin
     if (core.uiSettings.get('workspace:enabled') === false) {
+      // set default value for filtered nav links
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/plugin.ts:154]: Can we move this block to a private method with a meaningful name like `registerFilterLinksWhenFlagOff` so that others will know what is doing for this block from the method name?

File: src/core/public/core_system.ts (hunk 13 above)
Comment (wanglam) [reply src/core/public/core_system.ts]: How about check the uiSetting of workspace here? I think the OSD should have its default nav render logic. workspaces is just a plugin of OSD. We can't move the default nav generate logic to workspace plugin. @ruanyl What do you think about that?

Comment (wanglam) [reply src/core/public/core_system.ts]: To avoid unnecessary rendering, I have a new idea. We can find some way to call the `navLinks.getNavLinks$()` as default `navLinks$` property in `src/core/public/chrome/chrome_service.tsx`. We can update to pass `navLinks.getFilteredNavLinks$()` after the `filteredNavLinks` updated. So we don't need to subscribe `navLinks` update here.
